### PR TITLE
Source map location

### PR DESCRIPTION
**Pull request by sergey-shandar**

### CODE REVIEW

**Comment by sergey-shandar** [comment]
It's ready for review. It has some breaking changes in output:
- Now it's proper JSON instead of JSON fragments. 
- JSON References are pointing on original location
- A lot of fixes in a procedure to form a reference.

**Comment by sergey-shandar** [comment]
@sarangan12 @veronicagg @vladbarosan

File: openapi-diff/src/core/OpenApiDiff/Program.cs [hunk 1]
```diff
@@ -42,9 +44,19 @@ private static int Main(string[] args)
                 swaggerNew,
                 settings
             );
-            foreach (var msg in messages)
+
+            if (settings.JsonValidationMessages)
+            {
```
Comment (vladbarosan) [inline openapi-diff/src/core/OpenApiDiff/Program.cs]: is this for CLI usage ? and is the default false ?

File: openapi-diff/src/modeler/AutoRest.Swagger/Model/Response.cs [hunk 2]
```diff
@@ -46,6 +46,7 @@ public class OperationResponse : SwaggerBase<OperationResponse>
             var headers = Headers != null ? Headers : new Dictionary<string, Header>();
             var priorHeaders = priorResponse.Headers != null ? priorResponse.Headers : new Dictionary<string, Header>();
 
+            context.PushProperty("headers");
             foreach (var header in headers)
```
Comment (vladbarosan) [inline openapi-diff/src/modeler/AutoRest.Swagger/Model/Response.cs]: remind me whats the purpose of this context ?

Comment (sergey-shandar) [reply openapi-diff/src/modeler/AutoRest.Swagger/Model/Response.cs]: to keep path/reference to the object

File: openapi-diff/src/core/OpenApiDiff/Program.cs (hunk 1 above)
Comment (sergey-shandar) [reply openapi-diff/src/core/OpenApiDiff/Program.cs]: yeah, I need to remove completely. The flag is always true now.

File: openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs [hunk 3]
```diff
@@ -44,20 +58,25 @@ private static ObjectPath ParseRef(string s)
...
         private static IEnumerable<(JToken token, string name)> CompletePath(IEnumerable<Func<JToken, string>> p, JToken t)
-            => p.Select(v => {
-                var name = v(t);
-                t =
-                    t is JArray a ? int.TryParse(name, out var i) ? a[i] : null :
-                    t is JObject o ? FromObject(o, name) :
-                    null;
-                return (t, name);
-            });
+            => new[] { (t, "#") }
```
Comment (vladbarosan) [inline openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs]: seems that we will return something more here than before, `{t}#` extra , what would be a before and after sample return value

File: openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs [hunk 4]
```diff
@@ -66,10 +85,13 @@ public static string FileNameNorm(string fileName)
             => fileName.Replace("\\", "/");
 
         // https://tools.ietf.org/html/draft-ietf-appsawg-json-pointer-04
-        public string JsonPointer(IJsonDocument t) => CompletePath(t.Token)
-            .Select(v => v.name?.Replace("~", "~0")?.Replace("/", "~1"))
-            .Aggregate(FileNameNorm(t.FileName) + "#", (a, b) => a == null || b == null ? null : a + "/" + b);
-            
+        public string JsonPointer(IJsonDocument t)
+        {
+            var result = CompletePath(t.Token)
```
Comment (vladbarosan) [inline openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs]: what format is completePath that we need to make these replacements

File: openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs [hunk 5]
```diff
@@ -33,6 +34,19 @@ private ObjectPath(IEnumerable<Func<JToken, string>> path)
             return list == null || list.Count == 0 ? null : list[0].i.ToString();
         });
 
+        public static string PathName(string path)
```
Comment (vladbarosan) [inline openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs]: i know this is just an extracted method, but is there anything pointing to this regex ? ( i.e. is it documented anywhere )

File: openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs (hunk 3 above)
Comment (sergey-shandar) [reply openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs]: It's a root property, we needed.

File: openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs (hunk 4 above)
Comment (sergey-shandar) [reply openapi-diff/src/core/OpenApiDiff.Core/Logging/ObjectPath.cs]: Returns a sequence of property names, including the "#" string.

File: src/lib/validate.ts [hunk 6]
```diff
@@ -19,14 +19,14 @@ import * as OpenApiDiff from './validators/openApiDiff'
...
 export function compare(oldSwagger: string, newSwagger: string, options: OpenApiDiff.Options) {
-  if (!options) options = {};
+  if (!options) options = {}
 
-  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel;
-  log.filepath = options.logFilepath || log.filepath;
-  let openApiDiff = new OpenApiDiff.OpenApiDiff(options);
+  log.consoleLogLevel = options.consoleLogLevel || log.consoleLogLevel
+  log.filepath = options.logFilepath || log.filepath
+  let openApiDiff = new OpenApiDiff.OpenApiDiff(options)
```
Comment (vladbarosan) [inline src/lib/validate.ts]: const ?

File: src/lib/validate.ts [hunk 7]
```diff
@@ -49,11 +49,11 @@ export function compare(oldSwagger: string, newSwagger: string, options: OpenApi
...
 
-  return openApiDiff.compare(oldSwagger, newSwagger, oldTag, newTag);
-};
+  return openApiDiff.compare(oldSwagger, newSwagger, oldTag, newTag)
```
Comment (veronicagg) [inline src/lib/validate.ts]: Let's double check on this comparison, which was what @sarangan12 was referring to regarding the different inputs. Let's be consistent on the compare output for both functions, compare and compareTags

File: src/lib/validators/openApiDiff.ts [hunk 8]
```diff
@@ -2,19 +2,64 @@
...
+type Change = {
+  readonly location?: string
+  readonly path?: string
+  readonly ref?: string
+}
+
+type Message = {
+  readonly id: string
+  readonly code: string
+  readonly docurl: string
```
Comment (veronicagg) [inline src/lib/validators/openApiDiff.ts]: docUrl, capitalize Url

File: src/lib/validators/openApiDiff.ts [hunk 9]
```diff
@@ -2,19 +2,64 @@
...
+  readonly message: string
+  readonly mode: string
+  readonly type: string
+  readonly new: Change
```
Comment (veronicagg) [inline src/lib/validators/openApiDiff.ts]: should this be newLocation & oldLocation? and maybe Change -> ChangeLocation ?

File: src/lib/validators/openApiDiff.ts [hunk 10]
```diff
@@ -186,52 +229,46 @@ export class OpenApiDiff {
...
-          reject(err);
-        }
+    const cmd = `${this.dotNetPath()} ${this.openApiDiffDllPath()} -o ${oldSwagger} -n ${newSwagger}`
 
-        resolve(stdout);
-      });
-    });
+    log.debug(`Executing: "${cmd}"`)
+    const { stdout } = await exec(cmd, { encoding: 'utf8', maxBuffer: 1024 * 1024 * 64 })
+    const json = jsonParser.parse("", stdout) as Messages
```
Comment (veronicagg) [inline src/lib/validators/openApiDiff.ts]: should this be resultInJson? or something like that?

File: src/lib/validators/openApiDiff.ts [hunk 11]
```diff
@@ -186,52 +229,46 @@ export class OpenApiDiff {
...
 
-    return OpenApiDiffPromise;
+    const newJson = json.map(v => ({
```
Comment (veronicagg) [inline src/lib/validators/openApiDiff.ts]: name suggestion: updatedResultInJson? or resultWithOriginalLineNumbers?

Comment (veronicagg) [reply src/lib/validators/openApiDiff.ts]: name suggestion: v -> result? or validationMessage?

File: src/lib/validators/openApiDiff.ts [hunk 12]
```diff
@@ -2,19 +2,64 @@
...
+  readonly message: string
+  readonly mode: string
+  readonly type: string
+  readonly new: Change
+  readonly old: Change
+}
+
+type Messages = ReadonlyArray<Message>
+
+const updateChange = (change: Change, pf: ProcessedFile): Change => {
```
Comment (veronicagg) [inline src/lib/validators/openApiDiff.ts]: name suggestion: updateLocationWithOriginalLineNumbers?

File: src/lib/validators/openApiDiff.ts [hunk 13]
```diff
@@ -186,52 +229,46 @@ export class OpenApiDiff {
...
+      old: updateChange(v.old, oldSwaggerFile),
+    }))
+    return JSON.stringify(newJson)
   }
 }
 
 // Testing
-// let swagger = '/Users/vishrut/git-repos/azure-rest-api-specs/arm-network/2017-03-01/swagger/virtualNetworkGateway.json';
-// let obj = new OpenApiDiff();
+// let swagger = '/Users/vishrut/git-repos/azure-rest-api-specs/arm-network/2017-03-01/swagger/virtualNetworkGateway.json'
```
Comment (veronicagg) [inline src/lib/validators/openApiDiff.ts]: is this section useful? or should it be removed?

File: src/test/test.ts [hunk 14]
```diff
@@ -1,11 +1,80 @@
...
+        type: "Info"
+      }
+    ]
+    assert.deepStrictEqual(result, expected)
+  })
+  it("some-changes", async () => {
+    const diff = new index.OpenApiDiff({})
+    const oldFile = "src/test/some-changes/old.json"
+    const newFile = "src/test/some-changes/new.json"
+    const resultStr = await diff.compare(oldFile, newFile)
```
Comment (veronicagg) [inline src/test/test.ts]: if you wanted you could also add a test swapping the parameters, to check on deletion of path and location of "new" as non-existent - which should also flag a "breaking" change. I can't remember how the "breaking ones are being flagged.

File: src/lib/validate.ts (hunk 7 above)
Comment (sergey-shandar) [reply src/lib/validate.ts]: A new test has been added.

File: src/lib/validators/openApiDiff.ts (hunk 13 above)
Comment (sergey-shandar) [reply src/lib/validators/openApiDiff.ts]: Right. We don't need it anymore because we have unit tests now.

**Pull request merged**